### PR TITLE
Adds a test for and handles some edge cases for optimistic handover

### LINKED ISSUE #121: Bug and some more work for optimistic handover
- Need to handle the TODO in https://github.com/qkniep/alpenglow/blob/main/src/consensus/block_producer.rs#L200
- Need to handle the situation where the hash didn't change but the slot changed in https://github.com/qkniep/alpenglow/blob/main/src/consensus/block_producer.rs#L186

### PR DESCRIPTION
**Pull request by akhi3030**
closes #121 

- Adds a test case for `produce_block_parent_not_ready`
- handles a couple of left over edge cases in optimistic handover

### CODE REVIEW

File: src/consensus/block_producer.rs [hunk 1]
```diff
@@ -188,7 +188,7 @@ where
                         let start = Instant::now();
                         let (new_slot, new_hash) = res.unwrap();
                         let (mut payload, _maybe_duration) = produce_slice_future.await;
-                        if new_hash != parent_hash {
+                        if new_hash != parent_hash || new_slot != parent_slot {
```
Comment (akhi3030) [inline src/consensus/block_producer.rs]: do we need to check for slot?

File: src/consensus/block_producer.rs [hunk 2]
```diff
@@ -284,6 +294,7 @@ where
         payload: SlicePayload,
         maybe_duration: Option<Duration>,
     ) -> Result<Either<Hash, Duration>> {
+        println!("handling head = {header:?}");
```
Comment (akhi3030) [inline src/consensus/block_producer.rs]: remove prints

File: src/consensus/block_producer.rs [hunk 3]
```diff
@@ -649,4 +662,110 @@ mod tests {
...
+        let block_info = BlockInfo {
+            hash: [1; 32],
+            parent: (slot.prev(), [2; 32]),
+        };
+
+        let (first_slice_finished_tx, first_slice_finished_rx) = oneshot::channel();
+        let (start_second_slice_tx, start_second_slice_rx) = oneshot::channel();
+
+        let mut seq = Sequence::new();
+        let mut blockstore = MockBlockstore::new();
```
Comment (qkniep) [inline src/consensus/block_producer.rs:718]: This use of the MockBlockstore looks more complicated than doing the same for the actual blockstore. Or ist there some detail in this setup that I might be missing?

File: src/consensus/block_producer.rs [hunk 4]
```diff
@@ -212,9 +212,18 @@ where
             let is_last = slice_index.is_max() || maybe_duration.is_none();
             if is_last && !parent_ready_receiver.is_terminated() {
```
Comment (qkniep) [inline src/consensus/block_producer.rs:234]: This is only to handle the case where we produce all slices before seeing the ParentReady, right?

File: src/consensus/block_producer.rs (hunk 1 above)
Comment (qkniep) [reply src/consensus/block_producer.rs]: I would say, no, the first parent we reconstructed (thus we know it's valid), the second parent was reconstructed by good nodes (thus we know it's valid), a valid block has different hash if the slot is different.

Comment (akhi3030) [reply src/consensus/block_producer.rs]: OK, yes, that makes sense.  Thanks.

File: src/consensus/block_producer.rs (hunk 4 above)
Comment (akhi3030) [reply src/consensus/block_producer.rs:234]: Yes.

File: src/consensus/block_producer.rs (hunk 3 above)
Comment (akhi3030) [reply src/consensus/block_producer.rs:718]: No, I agree, this block of code is too complicated.  I wanted to checkpoint this PR and attempt to reproduce the test using the actual blockstore to see what that takes.

File: src/consensus/block_producer.rs (hunk 4 above)
Comment (qkniep) [reply src/consensus/block_producer.rs:234]: We might be able to do without this. Any rational block producer would want to prevent that from ever happening, instead waiting longer to produce their slices. Also, we might want to have a higher slice count limit for optimistic block production anyways. One generally reasonable strategy would be to wait with producing the last slice until the timeout runs out (so way after ParentReady is emitted). I will create a new issue for this.
